contrib/interconnect/ic_modules.c

```diff
@@ -146,20 +146,15 @@ _PG_init(void)
 				 errmsg("could not load interconnect outside process shared preload")));
     }
 
-    switch(Gp_interconnect_type) {
-        case INTERCONNECT_TYPE_TCP:
-            CurrentMotionIPCLayer = &tcp_ipc_layer;
-            break;
-        case INTERCONNECT_TYPE_UDPIFC:
-            CurrentMotionIPCLayer = &udpifc_ipc_layer;
-            break;
-        case INTERCONNECT_TYPE_PROXY:
-            CurrentMotionIPCLayer = &proxy_ipc_layer;
-            break;    
-        default:
-            ereport(ERROR,
-				(errcode_for_file_access(),
-				 errmsg("could not decide interconnect type")));
-    }
-    
+	IPCLayerImpls[CurrentIPCLayerImplNum++] = &tcp_ipc_layer;
+	IPCLayerImpls[CurrentIPCLayerImplNum++] = &udpifc_ipc_layer;
+	IPCLayerImpls[CurrentIPCLayerImplNum++] = &proxy_ipc_layer;
```

Comment:
1. should check the array boundary
2. ensure that the ic_type should be unique in the array

Reply:
fix:
1. check the array boundary;
2. ensure that the ic_type is unique in the array;